### PR TITLE
Add option to resolve IP from host if unique in GetURL

### PR DESCRIPTION
**Pull request by wkentaro**
GetURL can be used to compare a two ControllerClientInfo, but host can be either hostname or ip address. This can cause problems when comparison fails such as localhost vs. 127.0.0.1:

```
"http://user:password@localhost:123"
!= "http://user:password@127.0.0.1:123"
```

### CODE REVIEW

**Comment by wkentaro** [comment]
```
% ./resolve localhost
IP address of localhost is 127.0.0.1 
% ./resolve 127.0.0.1
IP address of 127.0.0.1 is 127.0.0.1 
% ./resolve google.com
IP address of google.com is 142.251.222.46
% ./resolve PF4PAEYL  
IP address of PF4PAEYL is 127.0.1.1
```

```cpp
#include <netdb.h>
#include <arpa/inet.h>
#include <iostream>
#include <cstring>
#include <set>
#include <string>

std::string ResolveIpFromHostnameIfUnique(const std::string& hostname) {
    struct addrinfo hints = {}, *addrs;

    memset(&hints, 0, sizeof(hints));
    hints.ai_family = AF_INET;  // IPv4 only
    hints.ai_socktype = SOCK_STREAM;

    int status = getaddrinfo(hostname.c_str(), nullptr, &hints, &addrs);
    if (status != 0) {
        std::cerr << "Hostname resolution failed: " << gai_strerror(status) << std::endl;
        return "";
    }

    std::set<std::string> uniqueIps;

    for (struct addrinfo* addr = addrs; addr != nullptr; addr = addr->ai_next) {
        char ipstr[INET_ADDRSTRLEN];
        auto* sockaddr = reinterpret_cast<struct sockaddr_in*>(addr->ai_addr);
        inet_ntop(AF_INET, &(sockaddr->sin_addr), ipstr, sizeof(ipstr));

        uniqueIps.insert(ipstr);
    }

    freeaddrinfo(addrs);

    if (uniqueIps.size() == 1) {
        return *uniqueIps.begin();
    }

    std::cerr << "Found " << uniqueIps.size() << " unique IP addresses" << std::endl;
    return "";
}

int main(int argc, char* argv[]) {
    if (argc != 2) {
        std::cerr << "Usage: " << argv[0] << " <hostname>" << std::endl;
        return 1;
    }

    const char* hostname = argv[1];
    std::string ip = ResolveIpFromHostnameIfUnique(hostname);
    std::cout << "IP address of " << hostname << " is " << ip << std::endl;

    return 0;
}
```
```

File: src/mujincontrollerclient.cpp [hunk 1]
```diff
@@ -89,7 +91,45 @@ bool ControllerClientInfo::operator==(const ControllerClientInfo &rhs) const
...
+
+    memset(&hints, 0, sizeof(hints));
+    hints.ai_family = AF_INET;  // IPv4 only
+    hints.ai_socktype = SOCK_STREAM;
+
+    int status = getaddrinfo(host.c_str(), nullptr, &hints, &addrs);
+    if (status != 0) {
+        std::stringstream ss;
+        ss << "host resolution failed for host " << host << ": " << gai_strerror(status);
+        MUJIN_LOG_ERROR(ss.str());
```
Comment (yoshikikanemoto) [inline src/mujincontrollerclient.cpp:106]: @wkentaro i think you can put `"host resolution failed for host " << host << ": " << gai_strerror(status)` as an argument.

File: src/mujincontrollerclient.cpp [hunk 2]
```diff
@@ -89,7 +91,45 @@ bool ControllerClientInfo::operator==(const ControllerClientInfo &rhs) const
...
+        return "";
+    }
+
+    std::set<std::string> uniqueIps;
+
+    for (struct addrinfo* addr = addrs; addr != nullptr; addr = addr->ai_next) {
+        char ipstr[INET_ADDRSTRLEN];
+        auto* sockaddr = reinterpret_cast<struct sockaddr_in*>(addr->ai_addr);
```
Comment (jschaare) [inline src/mujincontrollerclient.cpp:114]: Avoid using `auto` per the current style guide

File: src/mujincontrollerclient.cpp [hunk 3]
```diff
@@ -103,7 +143,13 @@ std::string ControllerClientInfo::GetURL(bool bIncludeNamePassword) const
         url += "@";
     }
 
-    url += host;
+    if (bResolveIpFromHostIfUnique) {
+        const std::string ip = ResolveIpFromHostIfUnique(host);
+        url += ip.empty() ? host : ip;
```
Comment (jschaare) [inline src/mujincontrollerclient.cpp:148]: If we're expecting this to guarantee uniqueness, is it correct behavior to fall back to using the `host`?

File: src/mujincontrollerclient.cpp [hunk 4]
```diff
@@ -15,6 +15,8 @@
 #include "controllerclientimpl.h"
 #include <boost/thread.hpp> // for sleep
 #include <boost/algorithm/string.hpp>
+#include <netdb.h>
+#include <arpa/inet.h>
```
Comment (jschaare) [inline src/mujincontrollerclient.cpp:19]: I think the implementation using `getaddrinfo` is fine, but should we consider an implementation using boost?

**Comment by wkentaro** [comment]
boost alternative:

image

```cpp
#include <boost/asio.hpp>
#include <iostream>
#include <set>

std::string ResolveIpFromHostIfUnique(const std::string& host)
{
    try {
        boost::asio::io_context io_context;
        boost::asio::ip::tcp::resolver resolver(io_context);
        boost::asio::ip::tcp::resolver::query query(boost::asio::ip::tcp::v4(), host, "http");
        boost::asio::ip::tcp::resolver::results_type results = resolver.resolve(query);

        std::set<std::string> uniqueIps;

        for (boost::asio::ip::tcp::resolver::results_type::iterator entry = results.begin(); entry != results.end(); ++entry) {
            uniqueIps.insert(entry->endpoint().address().to_string());
        }

        if (uniqueIps.size() == 1) {
            return *uniqueIps.begin();
        }

        std::cerr << "Error: Host " << host << " resolved to multiple IPs" << std::endl;
        for (const std::string& ip : uniqueIps) {
            std::cerr << "  " << ip << std::endl;
        }
        return "";
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return "";
    }
}


int main(int argc, char* argv[]) {
    if (argc != 2) {
        std::cerr << "Usage: " << argv[0] << " <host>" << std::endl;
        return 1;
    }

    const std::string host = argv[1];

    const std::string ip = ResolveIpFromHostIfUnique(host);
    if (!ip.empty()) {
        std::cout << ip << std::endl;
        return 0;
    }

    return 1;
}
```

File: src/mujincontrollerclient.cpp (hunk 4 above)
Comment (wkentaro) [reply src/mujincontrollerclient.cpp:19]: I tried boost version here: https://github.com/mujin/controllerclientcpp/pull/176#issuecomment-2513333935

Boost seems easy to read. Should we go with boost (not sure how to decide tbh)? cc @yoshikikanemoto

**Comment by wkentaro** [comment]
@jschaare @yoshikikanemoto I addressed all of the comments.

**Comment by yoshikikanemoto** [comment]
thanks

File: src/mujincontrollerclient.cpp (hunk 4 above)
Comment (yoshikikanemoto) [reply src/mujincontrollerclient.cpp:19]: i don't see much difference. to me, `getaddrinfo` is easier to read or get info.

**Pull request merged**